File: src/ctx/briefing.py
from __future__ import annotations

import json
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .gitinfo import git_info
from .map import build_map
from .pack import pack_query


def generate_briefing(root: Path, *, intent: str, focus: str | None, harness: str,
                       out: Path, force: bool = False) -> tuple[dict[str, Any], bool]:
    started = time.perf_counter()
    info = git_info(root)
    json_path = out / "briefing.json"
    md_path = out / "briefing.md"
    if json_path.exists() and not force and not info.dirty:
        try:
            previous = json.loads(json_path.read_text(encoding="utf-8"))
            if previous.get("sha") == info.sha:
                return {**previous, "skipped": True}, True
        except (OSError, json.JSONDecodeError):
            pass
    repo_map = build_map(root)
    packed = pack_query(focus, budget_tokens=2000, intent="explore", start=root) if focus else {
        "hits": [], "coverage": "complete", "hint": "Ajoutez --focus pour obtenir des preuves ciblées."
    }
    hits = [hit for hit in packed["hits"] if (root / hit["path"]).is_file()]
    stack = detect_stack(root)
    identity_evidence = _identity_evidence(root, repo_map)
    identity = {"one_liner": one_liner(root, stack), "stack": stack, "evidence": identity_evidence}
    hotspots = [
        {"path": hub["path"], "why": f"hub avec {hub['in_edges']} liens entrants", "risk": "couplage"}
        for hub in repo_map.get("hubs", []) if (root / hub["path"]).is_file()
    ][:7]
    touch = list(dict.fromkeys(hit["path"] for hit in hits if hit["kind"] in {"def", "call"}))[:10]
    tests = [hit["path"] for hit in hits if hit["kind"] == "test" or "test" in hit["path"].lower()]
    data: dict[str, Any] = {
        "schema": "ctx.briefing.v1", "repo": root.name, "sha": info.sha, "dirty": info.dirty,
        "intent": intent, "focus": focus, "generated_at": datetime.now(timezone.utc).isoformat(),
        "freshness_ms": int((time.perf_counter() - started) * 1000), "harness": harness,
        "identity": identity, "map": repo_map, "flows": [],
        "contracts": {"http": repo_map.get("router", []), "events": [], "tables": []},
        "hotspots": hotspots,
        "change_plan": {"touch": touch, "avoid": [], "tests": list(dict.fromkeys(tests)),
                        "blast_radius_files": len(touch), "depth": 2},
        "hits": hits, "coverage": packed["coverage"], "hint": packed.get("hint"),
    }
    data["freshness_ms"] = int((time.perf_counter() - started) * 1000)
    out.mkdir(parents=True, exist_ok=True)
    (out / "map.json").write_text(json.dumps(repo_map, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    json_path.write_text(json.dumps(data, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    md_path.write_text(render_markdown(data), encoding="utf-8")
    return data, False


def detect_stack(root: Path) -> list[str]:
    stack: list[str] = []
    manifests = {
        "pyproject.toml": "Python", "requirements.txt": "Python", "package.json": "JavaScript/TypeScript",
        "Cargo.toml": "Rust", "go.mod": "Go", "composer.json": "PHP",
    }
    for filename, label in manifests.items():
        if (root / filename).is_file() and label not in stack:
            stack.append(label)
    suffixes = {path.suffix for path in root.rglob("*") if path.is_file() and len(path.relative_to(root).parts) <= 3}
    if ".py" in suffixes and "Python" not in stack:
        stack.append("Python")
    if suffixes & {".ts", ".tsx", ".js"} and "JavaScript/TypeScript" not in stack:
        stack.append("JavaScript/TypeScript")
    for suffix, label in ((".go", "Go"), (".rs", "Rust"), (".php", "PHP")):
        if suffix in suffixes and label not in stack:
            stack.append(label)
    framework_markers = {
        "FastAPI": ("fastapi",), "Django": ("django",), "Flask": ("flask",), "Litestar": ("litestar",),
        "Next.js": ("\"next\"",), "Express": ("\"express\"",), "NestJS": ("@nestjs/",),
        "Fastify": ("\"fastify\"",), "Hono": ("\"hono\"",), "Vue": ("\"vue\"",), "Svelte": ("\"svelte\"",),
        "Gin": ("gin-gonic/gin",), "Echo": ("labstack/echo",), "Fiber": ("gofiber/fiber",), "Chi": ("go-chi/chi",),
        "Axum": ("axum",), "Actix Web": ("actix-web",), "Rocket": ("rocket",),
        "Laravel": ("laravel/framework",), "Symfony": ("symfony/framework-bundle",), "Slim": ("slim/slim",),
    }
    manifest_names = ("pyproject.toml", "requirements.txt", "package.json", "go.mod", "Cargo.toml", "composer.json")
    manifest_text = "\n".join((root / name).read_text(encoding="utf-8", errors="replace").lower() for name in manifest_names if (root / name).is_file())
    for framework, markers in framework_markers.items():
        if any(marker.lower() in manifest_text for marker in markers):
            stack.append(framework)
    return stack


def one_liner(root: Path, stack: list[str]) -> str:
    readme = next((p for p in (root / "README.md", root / "README.rst", root / "README.txt") if p.is_file()), None)
    if readme:
        paragraphs = [p.strip().replace("\n", " ") for p in readme.read_text(encoding="utf-8", errors="replace").split("\n\n") if p.strip()]
        if paragraphs:
            return paragraphs[0].lstrip("# ")[:300]
    suffix = f" ({', '.join(stack)})" if stack else ""
    return f"Dépôt {root.name}{suffix}"


def _identity_evidence(root: Path, repo_map: dict[str, Any]) -> str | None:
    for name in ("README.md", "README.rst", "README.txt", "pyproject.toml", "package.json", "go.mod", "Cargo.toml", "composer.json"):
        if (root / name).is_file():
            return f"{name}:1"
    packages = repo_map.get("packages", [])
    return packages[0].get("evidence") if packages else None


def render_markdown(data: dict[str, Any]) -> str:
    header = f"# Briefing {data['repo']} @ {data['sha']}   freshness: {data['freshness_ms']}ms   intent: {data['intent']}"
    identity = data["identity"]
    identity_line = identity["one_liner"]
    if identity.get("evidence"):
        identity_line += f" (`{identity['evidence']}`)"
    parts = [header, "", "## À quoi ça sert", "", identity_line]
    repo_map = data["map"]
    evidence = []
    if repo_map.get("entrypoints") or repo_map.get("packages"):
        parts += ["", "## Carte (où aller)", ""]
        for entry in repo_map.get("entrypoints", []):
            parts.append(f"- `{entry['path']}:{entry['line']}` — {entry['why']}")
            evidence.append(f"{entry['path']}:{entry['line']}")
        for package in repo_map.get("packages", [])[:10]:
            parts.append(f"- `{package['evidence']}` — {package['name']} ({package['files']} fichiers)")
            evidence.append(package["evidence"])
    if data.get("flows"):
        parts += ["", "## Flux (max 7)", ""] + [f"- {flow['name']}: {' -> '.join(flow['steps'])}" for flow in data["flows"][:7]]
    contracts = data.get("contracts", {})
    if any(contracts.values()):
        parts += ["", "## Contrats", ""]
        for item in contracts.get("http", []):
            parts.append(f"- `{item['path']}:{item['line']}` — {item['method']} {item['route']}")
            evidence.append(f"{item['path']}:{item['line']}")
    if data.get("hotspots"):
        parts += ["", "## Hotspots", ""]
        for item in data["hotspots"]:
            parts.append(f"- `{item['path']}:1` — {item['why']} ({item['risk']})")
            evidence.append(f"{item['path']}:1")
    if data["intent"] in {"change", "impact"}:
        plan = data["change_plan"]
        parts += ["", f"## Pour changer {data.get('focus') or 'le sujet'}", ""]
        if plan["touch"]:
            parts.extend(f"- Examiner `{path}:1`" for path in plan["touch"])
        else:
            parts.append("Aucun fichier cible établi; affiner `--focus`.")
    if data.get("hits"):
        parts += ["", "## Preuves", ""]
        for hit in data["hits"]:
            parts.append(f"- `{hit['path']}:{hit['start']}` — {hit['symbol']}: {hit['why']}")
    elif data.get("hint"):
        parts += ["", "## Preuves", "", data["hint"]]
    return "\n".join(parts) + "\n"

File: src/briefing.rs
use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Instant;

use anyhow::{Context, Result};
use serde_json::{Value, json};

use crate::gitinfo::git_info;
use crate::map::{RepositoryMap, build_map};
use crate::pack::pack_query;

pub fn generate_briefing(
    root: &Path,
    intent: &str,
    focus: Option<&str>,
    harness: &str,
    out: &Path,
    force: bool,
) -> Result<(Value, bool)> {
    let started = Instant::now();
    let info = git_info(root);
    let json_path = out.join("briefing.json");
    if json_path.is_file()
        && !force
        && !info.dirty
        && let Ok(mut previous) = read_json(&json_path)
        && previous.get("sha").and_then(Value::as_str) == Some(&info.sha)
    {
        previous["skipped"] = Value::Bool(true);
        return Ok((previous, true));
    }
    let repository_map = build_map(root)?;
    let packed = match focus {
        Some(query) => pack_query(query, 2_000, "explore", root)?,
        None => crate::model::Envelope::empty(
            "complete",
            Some("Ajoutez --focus pour obtenir des preuves ciblées.".to_owned()),
        ),
    };
    let hits = packed
        .hits
        .into_iter()
        .filter(|hit| root.join(&hit.path).is_file())
        .collect::<Vec<_>>();
    let stack = detect_stack(root)?;
    let identity_evidence = identity_evidence(root, &repository_map);
    let hotspots = repository_map
        .hubs
        .iter()
        .filter(|hub| root.join(&hub.path).is_file())
        .take(7)
        .map(|hub| {
            json!({
                "path": hub.path,
                "why": format!("hub avec {} liens entrants", hub.in_edges),
                "risk": "couplage"
            })
        })
        .collect::<Vec<_>>();
    let mut seen = HashSet::new();
    let touch = hits
        .iter()
        .filter(|hit| matches!(hit.kind.as_str(), "def" | "call"))
        .filter(|hit| seen.insert(hit.path.clone()))
        .map(|hit| hit.path.clone())
        .take(10)
        .collect::<Vec<_>>();
    let mut seen_tests = HashSet::new();
    let tests = hits
        .iter()
        .filter(|hit| hit.kind == "test" || hit.path.to_ascii_lowercase().contains("test"))
        .filter(|hit| seen_tests.insert(hit.path.clone()))
        .map(|hit| hit.path.clone())
        .collect::<Vec<_>>();
    let mut data = json!({
        "schema": "ctx.briefing.v1",
        "repo": root.file_name().and_then(|value| value.to_str()).unwrap_or("repo"),
        "sha": info.sha,
        "dirty": info.dirty,
        "intent": intent,
        "focus": focus,
        "generated_at": chrono::Utc::now().to_rfc3339(),
        "freshness_ms": started.elapsed().as_millis(),
        "harness": harness,
        "identity": {
            "one_liner": one_liner(root, &stack)?,
            "stack": stack,
            "evidence": identity_evidence,
        },
        "map": repository_map,
        "flows": [],
        "contracts": {
            "http": repository_map.router,
            "events": [],
            "tables": [],
        },
        "hotspots": hotspots,
        "change_plan": {
            "touch": touch,
            "avoid": [],
            "tests": tests,
            "blast_radius_files": touch.len(),
            "depth": 2,
        },
        "hits": hits,
        "coverage": packed.coverage,
        "hint": packed.hint,
    });
    data["freshness_ms"] = json!(started.elapsed().as_millis());
    fs::create_dir_all(out)?;
    fs::write(
        out.join("map.json"),
        format!("{}\n", serde_json::to_string_pretty(&repository_map)?),
    )?;
    fs::write(
        &json_path,
        format!("{}\n", serde_json::to_string_pretty(&data)?),
    )?;
    fs::write(out.join("briefing.md"), render_markdown(&data)?)?;
    Ok((data, false))
}

fn read_json(path: &Path) -> Result<Value> {
    serde_json::from_str(
        &fs::read_to_string(path)
            .with_context(|| format!("lecture impossible: {}", path.display()))?,
    )
    .with_context(|| format!("JSON invalide: {}", path.display()))
}

pub fn detect_stack(root: &Path) -> Result<Vec<String>> {
    let manifests = [
        ("pyproject.toml", "Python"),
        ("requirements.txt", "Python"),
        ("package.json", "JavaScript/TypeScript"),
        ("Cargo.toml", "Rust"),
        ("go.mod", "Go"),
        ("composer.json", "PHP"),
    ];
    let mut stack = Vec::new();
    let mut manifest_text = String::new();
    for (name, label) in manifests {
        let path = root.join(name);
        if path.is_file() {
            push_unique(&mut stack, label);
            manifest_text.push_str(
                &fs::read_to_string(path)
                    .unwrap_or_default()
                    .to_ascii_lowercase(),
            );
            manifest_text.push('\n');
        }
    }
    for entry in walkdir::WalkDir::new(root)
        .max_depth(3)
        .into_iter()
        .filter_map(Result::ok)
        .filter(|entry| entry.file_type().is_file())
    {
        match entry.path().extension().and_then(|value| value.to_str()) {
            Some("py" | "pyi" | "pyw") => push_unique(&mut stack, "Python"),
            Some("js" | "jsx" | "ts" | "tsx") => push_unique(&mut stack, "JavaScript/TypeScript"),
            Some("go") => push_unique(&mut stack, "Go"),
            Some("rs") => push_unique(&mut stack, "Rust"),
            Some("php" | "phtml") => push_unique(&mut stack, "PHP"),
            _ => {}
        }
    }
    let frameworks = [
        ("FastAPI", &["fastapi"][..]),
        ("Django", &["django"][..]),
        ("Flask", &["flask"][..]),
        ("Litestar", &["litestar"][..]),
        ("Next.js", &["\"next\""][..]),
        ("Express", &["\"express\""][..]),
        ("NestJS", &["@nestjs/"][..]),
        ("Fastify", &["\"fastify\""][..]),
        ("Hono", &["\"hono\""][..]),
        ("Vue", &["\"vue\""][..]),
        ("Svelte", &["\"svelte\""][..]),
        ("Gin", &["gin-gonic/gin"][..]),
        ("Echo", &["labstack/echo"][..]),
        ("Fiber", &["gofiber/fiber"][..]),
        ("Chi", &["go-chi/chi"][..]),
        ("Axum", &["axum"][..]),
        ("Actix Web", &["actix-web"][..]),
        ("Rocket", &["rocket"][..]),
        ("Laravel", &["laravel/framework"][..]),
        ("Symfony", &["symfony/framework-bundle"][..]),
        ("Slim", &["slim/slim"][..]),
    ];
    for (framework, markers) in frameworks {
        if markers.iter().any(|marker| manifest_text.contains(marker)) {
            push_unique(&mut stack, framework);
        }
    }
    Ok(stack)
}

fn push_unique(values: &mut Vec<String>, value: &str) {
    if !values.iter().any(|item| item == value) {
        values.push(value.to_owned());
    }
}

fn one_liner(root: &Path, stack: &[String]) -> Result<String> {
    for name in ["README.md", "README.rst", "README.txt"] {
        let path = root.join(name);
        if !path.is_file() {
            continue;
        }
        let text = fs::read_to_string(path)?;
        if let Some(paragraph) = text.split("\n\n").find(|value| !value.trim().is_empty()) {
            return Ok(paragraph
                .trim()
                .replace('\n', " ")
                .trim_start_matches('#')
                .trim()
                .chars()
                .take(300)
                .collect());
        }
    }
    let repo = root
        .file_name()
        .and_then(|value| value.to_str())
        .unwrap_or("repo");
    let suffix = if stack.is_empty() {
        String::new()
    } else {
        format!(" ({})", stack.join(", "))
    };
    Ok(format!("Dépôt {repo}{suffix}"))
}

fn identity_evidence(root: &Path, repository_map: &RepositoryMap) -> Option<String> {
    for name in [
        "README.md",
        "README.rst",
        "README.txt",
        "pyproject.toml",
        "package.json",
        "go.mod",
        "Cargo.toml",
        "composer.json",
    ] {
        if root.join(name).is_file() {
            return Some(format!("{name}:1"));
        }
    }
    repository_map
        .packages
        .first()
        .map(|package| package.evidence.clone())
}

pub fn render_markdown(data: &Value) -> Result<String> {
    let text = |path: &str| {
        data.pointer(path)
            .and_then(Value::as_str)
            .unwrap_or_default()
    };
    let mut lines = vec![
        format!(
            "# Briefing {} @ {}   freshness: {}ms   intent: {}",
            text("/repo"),
            text("/sha"),
            data["freshness_ms"],
            text("/intent")
        ),
        String::new(),
        "## À quoi ça sert".to_owned(),
        String::new(),
    ];
    let mut identity = text("/identity/one_liner").to_owned();
    if let Some(evidence) = data.pointer("/identity/evidence").and_then(Value::as_str) {
        identity.push_str(&format!(" (`{evidence}`)"));
    }
    lines.push(identity);
    let repository_map = &data["map"];
    let entries = repository_map["entrypoints"]
        .as_array()
        .cloned()
        .unwrap_or_default();
    let packages = repository_map["packages"]
        .as_array()
        .cloned()
        .unwrap_or_default();
    if !entries.is_empty() || !packages.is_empty() {
        lines.extend([
            String::new(),
            "## Carte (où aller)".to_owned(),
            String::new(),
        ]);
        for entry in entries {
            lines.push(format!(
                "- `{}:{}` — {}",
                entry["path"].as_str().unwrap_or_default(),
                entry["line"],
                entry["why"].as_str().unwrap_or_default()
            ));
        }
        for package in packages.into_iter().take(10) {
            lines.push(format!(
                "- `{}` — {} ({} fichiers)",
                package["evidence"].as_str().unwrap_or_default(),
                package["name"].as_str().unwrap_or_default(),
                package["files"]
            ));
        }
    }
    let contracts = data["contracts"]["http"]
        .as_array()
        .cloned()
        .unwrap_or_default();
    if !contracts.is_empty() {
        lines.extend([String::new(), "## Contrats".to_owned(), String::new()]);
        for item in contracts {
            lines.push(format!(
                "- `{}:{}` — {} {}",
                item["path"].as_str().unwrap_or_default(),
                item["line"],
                item["method"].as_str().unwrap_or_default(),
                item["route"].as_str().unwrap_or_default()
            ));
        }
    }
    let hotspots = data["hotspots"].as_array().cloned().unwrap_or_default();
    if !hotspots.is_empty() {
        lines.extend([String::new(), "## Hotspots".to_owned(), String::new()]);
        for item in hotspots {
            lines.push(format!(
                "- `{}:1` — {} ({})",
                item["path"].as_str().unwrap_or_default(),
                item["why"].as_str().unwrap_or_default(),
                item["risk"].as_str().unwrap_or_default()
            ));
        }
    }
    if matches!(text("/intent"), "change" | "impact") {
        lines.extend([
            String::new(),
            format!(
                "## Pour changer {}",
                data["focus"].as_str().unwrap_or("le sujet")
            ),
            String::new(),
        ]);
        let touch = data["change_plan"]["touch"]
            .as_array()
            .cloned()
            .unwrap_or_default();
        if touch.is_empty() {
            lines.push("Aucun fichier cible établi; affiner `--focus`.".to_owned());
        } else {
            for path in touch {
                lines.push(format!(
                    "- Examiner `{}:1`",
                    path.as_str().unwrap_or_default()
                ));
            }
        }
    }
    let hits = data["hits"].as_array().cloned().unwrap_or_default();
    if !hits.is_empty() {
        lines.extend([String::new(), "## Preuves".to_owned(), String::new()]);
        for hit in hits {
            lines.push(format!(
                "- `{}:{}` — {}: {}",
                hit["path"].as_str().unwrap_or_default(),
                hit["start"],
                hit["symbol"].as_str().unwrap_or("fichier"),
                hit["why"].as_str().unwrap_or_default()
            ));
        }
    } else if let Some(hint) = data["hint"].as_str() {
        lines.extend([
            String::new(),
            "## Preuves".to_owned(),
            String::new(),
            hint.to_owned(),
        ]);
    }
    Ok(format!("{}\n", lines.join("\n")))
}

pub fn default_out(root: &Path) -> PathBuf {
    crate::config::ctx_dir(root)
}
